### PR TITLE
IPNFT as ERC721.

### PR DESCRIPTION
**Pull request by elmariachi111**
all tests green.

### CODE REVIEW

File: src/crowdsale/CrowdSale.sol [hunk 1]
```diff
@@ -74,18 +74,16 @@ contract CrowdSale is ReentrancyGuard {
      * @return saleId
      */
     function startSale(Sale calldata sale) public returns (uint256 saleId) {
-        //todo: use 15 minutes
-        if (sale.closingTime < block.timestamp + 2 hours) {
+        if (sale.closingTime < block.timestamp) {
```
Comment (pashov) [inline src/crowdsale/CrowdSale.sol:77]: You can have an instantly closing sale now.
The `+ 2 hours` was a not bad solution

File: src/crowdsale/CrowdSale.sol [hunk 2]
```diff
@@ -74,18 +74,16 @@ contract CrowdSale is ReentrancyGuard {
...
 
         if (sale.auctionToken.decimals() != 18) {
             revert BadDecimals();
         }
 
-        //close to 0 cases lead to very confusing results
-        //todo: use 10^(decimals-2) to allow 0.01 ethers
-        if (sale.fundingGoal < 10 ** sale.biddingToken.decimals() || sale.salesAmount < 0.5 ether) {
+        //close to 0 cases lead to precision issues.Using 0.01 bidding tokens as minimium funding goal
+        if (sale.fundingGoal < 10 ** (sale.biddingToken.decimals() - 2) || sale.salesAmount < 0.5 ether) {
```
Comment (pashov) [inline src/crowdsale/CrowdSale.sol:86]: What if the `biddingToken` has 2 decimals (example is `EURS`)

File: src/crowdsale/CrowdSale.sol (hunk 1 above)
Comment (elmariachi111) [reply src/crowdsale/CrowdSale.sol:77]: yeah, we talked about that quite some time. Imo the team "wants" it, even though its pointless (it makes sales simpler to test in test & stage envs but imo we should rather preprocess / code configure this restricion). But what actually matters most here: I don't think that restricting the closing time breaks anything but user experience.

File: src/crowdsale/CrowdSale.sol (hunk 2 above)
Comment (elmariachi111) [reply src/crowdsale/CrowdSale.sol:86]: test case added

**Comment by elmariachi111** [comment]
uses _mint and making Reservable:mintReservation external to avoid reentrancies.

**Review by DorianWilhelm — APPROVED** [review]

**Pull request merged**